### PR TITLE
Add rate limit bans

### PR DESCRIPTION
**Pull request by dcadenas**
This PR adds rate-limit bans for abusive users. Works hand in hand with a commit living in a fork of strfry-policies [here](https://github.com/planetary-social/strfry-policies/commit/321be2096c88e6125cefb58755952af0882a3ed8).

More specific bans can be added later on top of this. But right now anyone abusing through the same ip will get banned.

The previous rate-limit will still be working, but on top of it, a new one, for cases of even more abuse is used, this one bans the ip for he period specified.

First need to wait to hit the rate limit:

```
╰─>$ nak event -k 3 ws://localhost:7777                                                                                                                                (base) 13:38:20
connecting to ws://localhost:7777... ok.
{"kind":3,"id":"2d2a51cdbce91960a7b97a732761e0b7f28737a9efe9850ab19ee413a51d8942","pubkey":"79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798","created_at":1726850301,"tags":[],"content":"","sig":"df1c6b11110711e35ae9b4a9c6212d49ebda426ada82d0e91e5625e2b8d1c81fdcdb331dd288501daae0c8ea3b40fc2cbb3b392ce3148615e082ac930c588932"}
publishing to ws://localhost:7777... success.

┬─[daniel@loca:~/c/n/nos-followers]─[13:38:21]─[G:main=]
╰─>$ nak event -k 3 ws://localhost:7777                                                                                                                                (base) 13:38:21
connecting to ws://localhost:7777... ok.
{"kind":3,"id":"f234551eefc789278530e913431ecee4264e82d2f0d4942e6cfa649155ed24a2","pubkey":"79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798","created_at":1726850303,"tags":[],"content":"","sig":"caffd2a9537ab330fff056270f0f3cfcad078c4f528f44c177e56425247aa8234eca5de0979212fc0357c0a0e4c01e888b5bd77aa22448ff0ec0a2e11ebf0bd6"}
publishing to ws://localhost:7777... failed: msg: rate-limited: too many requests
```

We now wait some time until the rate limit period is cleared, we are banned, but the user doesn't know:

```
┬─[daniel@loca:~/c/n/nos-followers]─[13:38:23]─[G:main=]
╰─>$ nak event -k 3 ws://localhost:7777                                                                                                                                (base) 13:38:23
connecting to ws://localhost:7777... ok.
{"kind":3,"id":"4b19cbba0eb4a70c6781f27d2301be683c13c81a9444938e97e2afa81b1b0ba9","pubkey":"79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798","created_at":1726850323,"tags":[],"content":"","sig":"3a546149c01aec6d364d5c7bc5fa42893b2146c70699291954e49a7ae326f30951e8eed9b2b3fcd941a227a671f3d0b225d32c54a7031c1ca208a951323a5666"}
publishing to ws://localhost:7777... success.
```

The event was rejected
```
┬─[daniel@loca:~/c/n/nos-followers]─[13:38:43]─[G:main=]
╰─>$ nak req -i "4b19cbba0eb4a70c6781f27d2301be683c13c81a9444938e97e2afa81b1b0ba9" ws://localhost:7777                                                                 (base) 13:38:43
connecting to ws://localhost:7777... ok.
```

After the ban period elapses, the ip can write again:

```
┬─[daniel@loca:~/c/n/nos-followers]─[13:39:01]─[G:main=]
╰─>$ nak event -k 3 ws://localhost:7777                                                                                                                                (base) 13:39:01
connecting to ws://localhost:7777... ok.
{"kind":3,"id":"2d504c0997f3ee51fef78d2c38eb6840c1400cfc5bd285626a7ea5166c5a6b72","pubkey":"79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798","created_at":1726850414,"tags":[],"content":"","sig":"704cdeb9cf43796a8e40b83f9e5f17c6be7c238c634f136419c9f56f6dbc007935c081f78e415dcde08cc45ece76fe5fed89771fc80ff828a9afbf515a399b30"}
publishing to ws://localhost:7777... success.

┬─[daniel@loca:~/c/n/nos-followers]─[13:40:14]─[G:main=]
╰─>$ nak req -i "2d504c0997f3ee51fef78d2c38eb6840c1400cfc5bd285626a7ea5166c5a6b72" ws://localhost:7777                                                                 (base) 13:40:14
connecting to ws://localhost:7777... ok.
{"kind":3,"id":"2d504c0997f3ee51fef78d2c38eb6840c1400cfc5bd285626a7ea5166c5a6b72","pubkey":"79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798","created_at":1726850414,"tags":[],"content":"","sig":"704cdeb9cf43796a8e40b83f9e5f17c6be7c238c634f136419c9f56f6dbc007935c081f78e415dcde08cc45ece76fe5fed89771fc80ff828a9afbf515a399b30"}
```

### CODE REVIEW

**Comment by dcadenas** [comment]
This is using the work done in a fork of strfry-policies that needs to be reviewed too: https://github.com/planetary-social/strfry-policies/pull/1

File: strfry/plugins/policies.ts [hunk 1]
```diff
@@ -7,19 +7,45 @@ import {
...
+  // Remember to leave the more loose rate limit policy at the top
+  [
+    rateLimitPolicy,
+    {
+      max: 20,
+      interval: one_minute,
+      ban_interval: two_days,
+      whitelist: [localhost, eventsIp],
+      // We use a different db url so that this limiter is not affected by the other limiters
+      databaseUrl: "sqlite:///tmp/banning-strfry-rate-limit-policy.sqlite3",
```
Comment (mplorentz) [inline strfry/plugins/policies.ts:36]: Won't this be cleared whenever the docker container is updated? I think it should be in a directory that is persisted.

**Review by mplorentz — APPROVED** [review]
If I'm understanding this correctly, this is essentially a more severe version of our previous rate limiting algorithm? I am simultaneously worried that this won't be enough to prevent ReplyGuy spam on low-traffic relays like dwebcamp.nos.social, and also may catch normal usage. But since this is supposed to be a quick hotfix let's deploy it and see what it does.

File: strfry/plugins/policies.ts (hunk 1 above)
Comment (dcadenas) [reply strfry/plugins/policies.ts:36]: Good point, I can use the db dir that is already a volume

**Pull request merged**